File: apps/api/src/execution/dailyBriefing.ts
import { prisma } from "../lib/prisma";
import { redis } from "../lib/redis";
import { getMarketContext, type MarketContextPayload } from "../services/marketContext";
import { computePerformance, type TradeStats } from "../services/performance";
import { tradedPairs } from "./dataFreshness";

const DAY_MS = 24 * 60 * 60 * 1000;

export interface BriefingTrade {
  symbol: string;
  direction: string;
  outcome: string | null;
  grade: string | null;
  rMultiple: number | null;
  lesson: string | null;
  closedAt: string | null;
}

export interface DailyBriefing {
  generatedAt: string;
  performance: {
    totalTrades: number;
    winRate: number;
    totalPnL: number;
    maxDrawdown: number;
    averageRR: number;
    expectancy: number;
    profitFactor: number;
    rExpectancy: number; // mean R-multiple from graded journals (the edge metric)
    gradedTrades: number;
  };
  gradeDistribution: Record<string, number>;
  recent24h: BriefingTrade[];
  upcomingHighImpactNews: { title: string; currency: string; scheduledAt: string }[];
  topLessons: string[];
  // Phase 3 market-context agent: the AI's morning read (bias, levels, risks)
  // for each traded series. Empty when the AI service is unreachable.
  marketContext: MarketContextPayload[];
}

/** Market context for every traded series — best-effort, never throws. */
export async function collectMarketContext(): Promise<MarketContextPayload[]> {
  const out: MarketContextPayload[] = [];
  try {
    for (const { symbol, timeframe } of await tradedPairs()) {
      try {
        out.push(await getMarketContext(symbol, timeframe));
      } catch (err) {
        console.warn(
          `[dailyBriefing] market context unavailable for ${symbol}/${timeframe}: ${err instanceof Error ? err.message : err}`,
        );
      }
    }
  } catch (err) {
    console.warn(
      `[dailyBriefing] market context skipped: ${err instanceof Error ? err.message : err}`,
    );
  }
  return out;
}

/**
 * The agent's morning routine: summarize how it has been trading before the day
 * starts — performance + expectancy, the grades/lessons from recently closed
 * trades, and the high-impact news ahead. Persists `briefing:latest` to Redis
 * for the dashboard. Best-effort: never throws into the startup path.
 */
export async function runDailyBriefing(): Promise<DailyBriefing> {
  const now = new Date();
  const since24h = new Date(now.getTime() - DAY_MS);

  const closed = await prisma.trade.findMany({
    where: { status: "CLOSED" },
    include: { signal: true, journals: { orderBy: { createdAt: "desc" }, take: 1 } },
    orderBy: { closedAt: "desc" },
  });

  const stats: TradeStats[] = closed.map((t) => ({
    entryPrice: Number(t.entryPrice),
    exitPrice: t.exitPrice ? Number(t.exitPrice) : null,
    profitLoss: t.profitLoss ? Number(t.profitLoss) : null,
    direction: t.signal.direction as "LONG" | "SHORT",
    stopLoss: Number(t.signal.stopLoss),
  }));
  const perf = computePerformance(stats);

  // R-expectancy + grade distribution from the per-trade journal reviews.
  const rmults: number[] = [];
  const gradeDistribution: Record<string, number> = {};
  for (const t of closed) {
    const j = t.journals[0];
    if (!j) continue;
    if (j.rMultiple != null) rmults.push(Number(j.rMultiple));
    if (j.grade) gradeDistribution[j.grade] = (gradeDistribution[j.grade] ?? 0) + 1;
  }
  const rExpectancy = rmults.length ? rmults.reduce((a, b) => a + b, 0) / rmults.length : 0;

  const recent24h: BriefingTrade[] = closed
    .filter((t) => t.closedAt && t.closedAt >= since24h)
    .map((t) => {
      const j = t.journals[0];
      return {
        symbol: t.signal.symbol,
        direction: t.signal.direction,
        outcome: j?.outcome ?? null,
        grade: j?.grade ?? null,
        rMultiple: j?.rMultiple != null ? Number(j.rMultiple) : null,
        lesson: j?.lesson ?? null,
        closedAt: t.closedAt?.toISOString() ?? null,
      };
    });

  const news = await prisma.newsEvent.findMany({
    where: { scheduledAt: { gt: now, lt: new Date(now.getTime() + DAY_MS) }, impact: "HIGH" },
    orderBy: { scheduledAt: "asc" },
    take: 10,
  });

  const topLessons = closed
    .map((t) => t.journals[0]?.lesson)
    .filter((l): l is string => Boolean(l))
    .slice(0, 5);

  const marketContext = await collectMarketContext();

  const briefing: DailyBriefing = {
    generatedAt: now.toISOString(),
    performance: {
      ...perf,
      rExpectancy: Math.round(rExpectancy * 1000) / 1000,
      gradedTrades: rmults.length,
    },
    gradeDistribution,
    recent24h,
    upcomingHighImpactNews: news.map((n) => ({
      title: n.title,
      currency: n.currency,
      scheduledAt: n.scheduledAt.toISOString(),
    })),
    topLessons,
    marketContext,
  };

  try {
    if (redis.status === "ready") {
      await redis.set("briefing:latest", JSON.stringify(briefing));
    }
  } catch {
    // dashboard cache is best-effort
  }

  console.log(
    `[dailyBriefing] ${now.toISOString()} trades=${perf.totalTrades} win=${perf.winRate}% ` +
      `expectancy=$${perf.expectancy} R=${briefing.performance.rExpectancy} PF=${perf.profitFactor} ` +
      `recent24h=${recent24h.length} highImpactNews=${news.length} ` +
      `marketContext=${marketContext.map((m) => `${m.symbol}/${m.timeframe}:${m.bias}`).join(",") || "none"}`,
  );
  return briefing;
}

File: services/backend/app/domain/execution/daily_briefing.py
"""The agent's morning routine — port of `execution/dailyBriefing.ts`.

Summarizes how the desk has been trading before the day starts: performance +
expectancy, the grades/lessons from recently closed trades, and the high-impact
news ahead. Persists `briefing:latest` to Redis for the dashboard. Best-effort:
never raises into the startup path.
"""
from __future__ import annotations

import json
import logging
from datetime import timedelta
from typing import Any

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ...core.serialization import iso, num_or_none
from ...db.enums import Impact, TradeStatus
from ...db.models import NewsEvent, Trade
from ...db.redis_client import cache_set
from ...jobs.clock import naive_utcnow
from ..performance.metrics import TradeStats, compute_performance
from .data_freshness import traded_pairs

log = logging.getLogger("backend.briefing")

DAY = timedelta(days=1)


async def collect_market_context(session: AsyncSession) -> list[dict[str, Any]]:
    """Market context for every traded series — best-effort, never raises."""
    from ..market_context import get_market_context

    out: list[dict[str, Any]] = []
    try:
        for pair in await traded_pairs(session):
            try:
                out.append(
                    await get_market_context(session, pair["symbol"], pair["timeframe"])
                )
            except Exception as exc:  # noqa: BLE001
                log.warning(
                    "[dailyBriefing] market context unavailable for %s/%s: %s",
                    pair["symbol"],
                    pair["timeframe"],
                    exc,
                )
    except Exception as exc:  # noqa: BLE001
        log.warning("[dailyBriefing] market context skipped: %s", exc)
    return out


async def run_daily_briefing(session: AsyncSession) -> dict[str, Any]:
    """Build the briefing, cache it in Redis, and log the one-line summary."""
    now = naive_utcnow()
    since_24h = now - DAY

    closed = (
        (
            await session.execute(
                select(Trade)
                .where(Trade.status == TradeStatus.CLOSED)
                .order_by(Trade.closedAt.desc())
            )
        )
        .scalars()
        .all()
    )

    stats = [
        TradeStats(
            entryPrice=float(t.entryPrice),
            exitPrice=num_or_none(t.exitPrice),
            profitLoss=num_or_none(t.profitLoss),
            direction=t.signal.direction.value,
            stopLoss=float(t.signal.stopLoss),
        )
        for t in closed
    ]
    perf = compute_performance(stats)

    # R-expectancy + grade distribution from the per-trade journal reviews.
    rmults: list[float] = []
    grade_distribution: dict[str, int] = {}
    latest_journal: dict[str, Any] = {}
    for t in closed:
        journal = max(t.journals, key=lambda j: j.createdAt) if t.journals else None
        latest_journal[t.id] = journal
        if journal is None:
            continue
        if journal.rMultiple is not None:
            rmults.append(float(journal.rMultiple))
        if journal.grade:
            grade_distribution[journal.grade] = grade_distribution.get(journal.grade, 0) + 1
    r_expectancy = sum(rmults) / len(rmults) if rmults else 0.0

    recent_24h = []
    for t in closed:
        if t.closedAt is None or t.closedAt < since_24h:
            continue
        journal = latest_journal.get(t.id)
        recent_24h.append(
            {
                "symbol": t.signal.symbol,
                "direction": t.signal.direction.value,
                "outcome": getattr(journal, "outcome", None),
                "grade": getattr(journal, "grade", None),
                "rMultiple": (
                    float(journal.rMultiple)
                    if journal is not None and journal.rMultiple is not None
                    else None
                ),
                "lesson": getattr(journal, "lesson", None),
                "closedAt": iso(t.closedAt),
            }
        )

    news = (
        (
            await session.execute(
                select(NewsEvent)
                .where(
                    NewsEvent.scheduledAt > now,
                    NewsEvent.scheduledAt < now + DAY,
                    NewsEvent.impact == Impact.HIGH,
                )
                .order_by(NewsEvent.scheduledAt.asc())
                .limit(10)
            )
        )
        .scalars()
        .all()
    )

    top_lessons = [
        journal.lesson
        for journal in (latest_journal.get(t.id) for t in closed)
        if journal is not None and journal.lesson
    ][:5]

    market_context = await collect_market_context(session)

    briefing = {
        "generatedAt": iso(now),
        "performance": {
            **perf,
            "rExpectancy": round(r_expectancy, 3),
            "gradedTrades": len(rmults),
        },
        "gradeDistribution": grade_distribution,
        "recent24h": recent_24h,
        "upcomingHighImpactNews": [
            {"title": n.title, "currency": n.currency, "scheduledAt": iso(n.scheduledAt)}
            for n in news
        ],
        "topLessons": top_lessons,
        "marketContext": market_context,
    }

    await cache_set("briefing:latest", json.dumps(briefing))

    context_summary = (
        ",".join(f"{m['symbol']}/{m['timeframe']}:{m['bias']}" for m in market_context)
        or "none"
    )
    log.info(
        "[dailyBriefing] %s trades=%s win=%s%% expectancy=$%s R=%s PF=%s recent24h=%d "
        "highImpactNews=%d marketContext=%s",
        iso(now),
        perf["totalTrades"],
        perf["winRate"],
        perf["expectancy"],
        briefing["performance"]["rExpectancy"],
        perf["profitFactor"],
        len(recent_24h),
        len(news),
        context_summary,
    )
    return briefing
